java/src/org/openqa/selenium/grid/web/JarFileResource.java

```diff
@@ -32,9 +32,9 @@
 
 public class JarFileResource implements Resource {
```

Comment:
Enhance this class to include a new method that looks like below

```java
  protected JarFileResource newInstance(JarFile jarFile, String entryName, String prefix) {
    return new JarFileResource(jarFile, entryName, prefix);
  }

```

And this method would be used as below in this class

```java
  @Override
  public Optional<Resource> get(String path) {
    Require.nonNull("Path", path);

    if (!isDirectory()) {
      return Optional.empty();
    }

    String name = stripPrefix + stripLeadingSlash(path);

    ZipEntry entry = jarFile.getEntry(name);
    return Optional.ofNullable(entry).map(e -> newInstance(jarFile, entry.getName(), name));
  }
```

```java
  @Override
  public Set<Resource> list() {
    if (!isDirectory()) {
      return ImmutableSet.of();
    }

    String prefix = entryName.endsWith("/") ? entryName : entryName + "/";
    int count = prefix.split("/").length + 1;

    return jarFile.stream()
      .filter(e -> e.getName().startsWith(prefix))
      .filter(e -> !e.getName().equals(entryName))
      .filter(e -> !e.getName().equals(prefix))
      .filter(e -> e.getName().split("/").length == count)
      .map(e -> newInstance(jarFile, e.getName(), prefix))
      .collect(ImmutableSet.toImmutableSet());
  }
```

Reply:
Done the changes.